### PR TITLE
fix: Remove WeakList in favor of built-ins

### LINKED ISSUE #1506: Exceptions related to weakref in pyhf 0.6.2
# Description

With `pyhf` 0.6.2 I see a long trace of `weakref`-related exceptions at the end of a program using certain parts of `pyhf`. I have not extensively tested which parts cause this behavior, but have so far observed it with `pyhf.simplemodels.correlated_background` / `pyhf.simplemodels.uncorrelated_background`. I can also confirm that other functionality like `pyhf.set_backend("numpy")` does not trigger it.

# Expected Behavior

No stack of `weakref` exceptions after code finishes running.

# Actual Behavior

Large stack of `weakref`-related exceptions.

# Steps to Reproduce

Run a simple file using `pyhf.simplemodels.uncorrelated_background`, the example runs in a `python:3.8-slim` container with Python 3.8.11. I see the same locally in a conda environment with Python 3.8.10 on Mac OS. With `pyhf` 0.6.2 the stack of exceptions appears. When installing `pyhf==0.6.1` and using `pyhf.simplemodels.hepdata_like`, I do not see any exceptions. When running via `python -m pdb test.py` locally to debug, I do no longer see the exception. The current `pyhf` master also shows the same issue.

```python
$ cat test.py
import pyhf

model = pyhf.simplemodels.uncorrelated_background(
    signal=[5.0], bkg=[10.0], bkg_uncertainty=[2.0]
)
```
```console
$ docker run -it --rm -v$PWD:/test python:3.8-slim bash
root@8fd8a5f229c8:/# pip install pyhf
Collecting pyhf
  Downloading pyhf-0.6.2-py3-none-any.whl (140 kB)
     |████████████████████████████████| 140 kB 3.5 MB/s
Collecting pyyaml>=5.1
  Downloading PyYAML-5.4.1-cp38-cp38-manylinux1_x86_64.whl (662 kB)
     |████████████████████████████████| 662 kB 5.7 MB/s
Collecting scipy>=1.4.1
  Downloading scipy-1.7.0-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.whl (28.4 MB)
     |████████████████████████████████| 28.4 MB 16.4 MB/s
Collecting jsonpatch>=1.15
  Downloading jsonpatch-1.32-py2.py3-none-any.whl (12 kB)
Collecting tqdm>=4.56.0
  Downloading tqdm-4.61.1-py2.py3-none-any.whl (75 kB)
     |████████████████████████████████| 75 kB 2.9 MB/s
Collecting click>=7.0
  Downloading click-8.0.1-py3-none-any.whl (97 kB)
     |████████████████████████████████| 97 kB 6.2 MB/s
Collecting jsonschema>=3.0.0
  Downloading jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
     |████████████████████████████████| 56 kB 6.3 MB/s
Collecting jsonpointer>=1.9
  Downloading jsonpointer-2.1-py2.py3-none-any.whl (7.4 kB)
Requirement already satisfied: setuptools in /usr/local/lib/python3.8/site-packages (from jsonschema>=3.0.0->pyhf) (57.0.0)
Collecting six>=1.11.0
  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting attrs>=17.4.0
  Downloading attrs-21.2.0-py2.py3-none-any.whl (53 kB)
     |████████████████████████████████| 53 kB 4.1 MB/s
Collecting pyrsistent>=0.14.0
  Downloading pyrsistent-0.18.0-cp38-cp38-manylinux1_x86_64.whl (118 kB)
     |████████████████████████████████| 118 kB 18.9 MB/s
Collecting numpy<1.23.0,>=1.16.5
  Downloading numpy-1.21.0-cp38-cp38-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (15.7 MB)
     |████████████████████████████████| 15.7 MB 18.2 MB/s
Installing collected packages: six, pyrsistent, numpy, jsonpointer, attrs, tqdm, scipy, pyyaml, jsonschema, jsonpatch, click, pyhf
Successfully installed attrs-21.2.0 click-8.0.1 jsonpatch-1.32 jsonpointer-2.1 jsonschema-3.2.0 numpy-1.21.0 pyhf-0.6.2 pyrsistent-0.18.0 pyyaml-5.4.1 scipy-1.7.0 six-1.16.0 tqdm-4.61.1
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
root@8fd8a5f229c8:/# python test/test.py
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf03a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf01f0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0280>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0430>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf00d0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0160>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0670>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf05e0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf04c0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0550>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0820>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0790>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0700>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf09d0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0940>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf08b0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0c10>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0b80>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0a60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0af0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0dc0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0d30>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0ca0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0f70>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0ee0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dcf0e50>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dc821f0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dc82160>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dc82040>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dc820d0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dc823a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dc82280>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dc82310>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dc82430>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
Exception ignored in: <function WeakMethod.__new__.<locals>._cb at 0x7f7e8dc824c0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/weakref.py", line 58, in _cb
  File "/usr/local/lib/python3.8/weakref.py", line 74, in __eq__
TypeError: isinstance() arg 2 must be a type or tuple of types
```

# Checklist

- [X] Run `git fetch` to get the most up to date version of `master`
- [X] Searched through existing Issues to confirm this is not a duplicate issue
- [X] Filled out the Description, Expected Behavior, Actual Behavior, and Steps to Reproduce sections above or have edited/removed them in a way that fully describes the issue

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Resolves #1506

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR
```
* Rewrite the events code to handle weakrefs more properly
   - This is needed to work with Click v8.0+
* Fix up tests that were incorrectly written due to API changes in unittest.mock
```

### CODE REVIEW

File: src/pyhf/events.py [hunk 1]
```diff
@@ -24,19 +23,41 @@ def noop(*args, **kwargs):
     pass
 
 
-class WeakList(list):
-    def append(self, item):
-        list.append(self, weakref.WeakMethod(item, self.remove))
```
Comment (matthewfeickert) [inline src/pyhf/events.py:29]: I need to teach myself about [`weakref.WeakMethod`](https://docs.python.org/3/library/weakref.html#weakref.WeakMethod) but I'm not clear on why we won't need to do a `self.remove` in the `Callables` append method.

**Comment by matthewfeickert** [comment]
I think something is going wrong with the `conftest` as I can trigger the errors seen in `doctest` locally with

```console
$ python -m pytest -x -r sx --ignore tests/benchmarks/ --ignore tests/contrib --ignore tests/test_notebooks.py
```
```pytb
src/pyhf/infer/utils.py ..                                                                                                                                                                          [  3%]
src/pyhf/tensor/jax_backend.py F

================================================================================================ FAILURES =================================================================================================
_________________________________________________________________________ [doctest] pyhf.tensor.jax_backend.jax_backend.astensor __________________________________________________________________________
201 
202         Convert to a JAX ndarray.
203 
204         Example:
205 
206             >>> import pyhf
207             >>> pyhf.set_backend("jax")
UNEXPECTED EXCEPTION: TypeError("'NoneType' object is not callable")
Traceback (most recent call last):
  File "/home/feickert/.pyenv/versions/3.8.6/lib/python3.8/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest pyhf.tensor.jax_backend.jax_backend.astensor[1]>", line 1, in <module>
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/events.py", line 122, in register_wrapper
    result = func(*args, **kwargs)
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/__init__.py", line 147, in set_backend
    events.trigger("tensorlib_changed")()
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/events.py", line 53, in __call__
    func()(*args, **kwargs)
TypeError: 'NoneType' object is not callable
/home/feickert/Code/GitHub/pyhf/src/pyhf/tensor/jax_backend.py:207: UnexpectedException
```

but 

```console
$ pytest -x --doctest-modules src/pyhf/infer/utils.py src/pyhf/tensor/numpy_backend.py src/pyhf/tensor/jax_backend.py
```

runs fine

File: src/pyhf/events.py (hunk 1 above)
Comment (kratsg) [reply src/pyhf/events.py:29]: We could, but relying on the callback of something to be deleted is a bit trickier when iterating over loops where something gets deleted during loop iteration. It's generally going to be easier to flush things out every time you access something in the weak list.

**Comment by kratsg** [comment]
Just putting a note that coverage needs to be fixed, but this is confirmed to work by @alexander-held .

**Comment by matthewfeickert** [comment]
> Just putting a note that coverage needs to be fixed, but this is confirmed to work by @alexander-held .

Once this is in and we're happy with it working let's cut a `v0.6.3` release so that this won't keep hitting `cabinetry` and the rest of our users. :+1:

File: src/pyhf/events.py [hunk 2]
```diff
@@ -24,19 +23,66 @@ def noop(*args, **kwargs):
...
+        """
+        Append a new bound method as a callback to the list of callables.
+        """
+        try:
+            # methods
+            callback_ref = weakref.ref(callback.__func__), weakref.ref(
+                callback.__self__
+            )
+        except AttributeError:
+            callback_ref = weakref.ref(callback), None
```
Comment (matthewfeickert) [inline src/pyhf/events.py:48]: I'm sure that I'm missing the obvious here, but if this isn't already being hit anywhere in CI then what are the types of events that _would_ trigger this?

File: src/pyhf/events.py [hunk 3]
```diff
@@ -24,19 +23,66 @@ def noop(*args, **kwargs):
...
+        self._callbacks.append(callback_ref)
```
Comment (matthewfeickert) [inline src/pyhf/events.py:49]:
```suggestion
        self._callbacks.append(
            (weakref.ref(callback.__func__), weakref.ref(callback.__self__))
        )
```

File: src/pyhf/events.py [hunk 4]
```diff
@@ -24,19 +23,66 @@ def noop(*args, **kwargs):
...
+        for func, arg in self.callbacks:
             # weakref: needs to be de-ref'd first before calling
-            func()(*args, **kwargs)
+            if arg is not None:
+                func()(arg(), *args, **kwargs)
+            else:
+                func()(*args, **kwargs)
+
+    def __iter__(self):
+        return iter(self.callbacks)
```
Comment (matthewfeickert) [inline src/pyhf/events.py:76]:
```suggestion
    def __call__(self, *args, **kwargs):
        for func, arg in self.callbacks:
            func()(arg(), *args, **kwargs)
```

**Review by matthewfeickert — COMMENTED** [review]
I'm sure that I'm missing the obvious here, but if this isn't already being hit anywhere in CI then what are the types of events that _would_ trigger this?

File: src/pyhf/events.py (hunk 3 above)
Comment (matthewfeickert) [reply src/pyhf/events.py:49]: Ah a very easy counter example that I missed is the following commented out example in the file:

```python
>>> import pyhf
>>> @pyhf.events.subscribe('myevent')
... def test(a,b):
...     print(a+b)
... 
>>> pyhf.events.trigger("myevent")(1,2)
3
```

If you tried it with things removed like I suggested above you'd get the following error

```python
>>> import pyhf
>>> @pyhf.events.subscribe('myevent')
... def test(a,b):
...     print(a+b)
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/events.py", line 91, in __decorator
    __events.setdefault(event, Callables()).append(func)
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/events.py", line 43, in append
    (weakref.ref(callback.__func__), weakref.ref(callback.__self__))
AttributeError: 'function' object has no attribute '__func__'
```

My bad.

**Review by kratsg — APPROVED** [review]

**Review by matthewfeickert — COMMENTED** [review]
Highest of fives @kratsg. This is great. :)

Also thanks for these tests as this better shows me how to do things with `Mock`. :+1:

**Pull request merged**